"# Here is the journey recording the way of someone like me, growing from a little potato into a little bigger potato."							

"Questions,conversations,characters,numbers"
#Answers, captions, data, short comments
code

#---------2017/6/6------------							
"I have difficulty to reproduct the code in monocle manual"							
HSMM<-estimateDispersions(HSMM)							
disp_table <- dispersionTable(HSMM)							
							
"I updata OS, R and R studio, didn't work."							
#"Finally I realize Ctrapnell mentioned this in his github that:							
# ctrapnell commented on Jun 7 2016							
# estimateDispersions only makes sense right now for negbinomial and negbinomial.size expression families."							
							
#https://github.com/cole-trapnell-lab/monocle-release/issues/5							
							
#"So I can changed expressionFamily=negbinomial.size() and add  lowerDetectionLimit=0.5 while build newCellData							
#It works."							
							
#---------2017/6/9------------							
						
#ggplot legend tricks:	
theme()
text = element_text(size=40)     #larger text including legend title							
legend.text = element_text(size=30) #larger legend text							
legend.key.size = unit(2, "cm") #legend height and width
legend.key.width = unit(2, "cm") #legend width
legend.key.height = unit(2, "cm") #legend height
legend.position="none" #remove legend
							
unit(2, "cm")==grid::unit(1, "in")
							
legend.justification = 'right'						
legend.position=c(0.3,0.82) # change legend position, c() can be negative!"							
guides(colour = guide_legend(override.aes = list(size=10))) #larger legend label							
							
#To control the order and color, change the factor levels to (Epi,Endo, Immune)"
pData(HSMM_3)$CellType <-factor(pData(HSMM_3)$CellType,levels=c("KRT19+ epithelial Cells",
"Stromal+\nEndothelial+\nFibroblasts Cells",
"Immune Cells"))

#CellType and Cluster can be generated by runing reduceDimension()"

#Monocle sequence:
#1)reduceDimension
#2)clusterCells							
#3)plot_cell_clusters							
							
#1)reduceDimension							
#2)orderCells							
#3)plot_cell_trajectory							
							
#---------2017/6/10------------							
#host a Shiny Server/RStudio Server yourself							
#http://deanattali.com/2015/05/09/setup-rstudio-shiny-server-digital-ocean/							
							
#Step 3: Log in to your very own shiny new server							
ssh -l root 67.205.148.56  not  ssh 67.205.148.56 that mentioned in tutoiral							that mentioned in 
su - yah2014							
							
#6.1: Important note re: installing R packages							
sudo su - -c ""R -e \""install.packages('reshape', repos='http://cran.rstudio.com/')\							
sudo su - -c ""R -e \""install.packages('pheatmap', repos='http://cran.rstudio.com/')\						
sudo su - -c ""R -e \""biocLite('monocle', repos='https://bioconductor.org/biocLite.R/')\
'didnt work'

#"log in root and install inside R"							
							
#Shiny Server 32-bit Installation from Source							
#https://www.r-bloggers.com/installing-and-running-shiny-server-from-source-on-32-bit-ubuntu/


#---------2017/6/12 Completed Monocle manual------------


#Monocle manual mistake
'5.2 Finding genes that distinguish cell type or state'
#gene UBC should be the TBP

#---------2017/6/13 Build Shiny------------
rstudio.com/products/shiny/shiny-user-showcase/
bit.ly/shiny-quickstart-2
actionButton(inputId = "go", label = "Update")

#How to save and load S4 dataset
saveRDS(HSMM_epi_3,file = "HSMM_epi_3", ascii = T)
readRDS("HSMM_epi")

#How to generate plot in app
#when as.character turn character into number, use unlist
as.character(unlist(HSMM_ordering_genes))

#How to storage data in dropbox remotely
library(rdrop2)
token <- readRDS("droptoken.rds")
drop_acc(dtoken = token)
outputDir<-"/public/olivier/r/single_cell/rshiny/shinyapps/"
HSMM_ordering_genes <-drop_read_csv(paste0(outputDir,"HSMM_ordering_genes.csv"), dtoken=token)
HSMM_ordering_genes<-as.character(unlist(HSMM_ordering_genes))

drop_read_RDS <- function(file, dest = tempdir(), dtoken = get_dropbox_token(), ...) {
localfile = paste0(dest, "/", basename(file))
drop_get(file, local_file = localfile, overwrite = TRUE, dtoken = dtoken)
readRDS(localfile, ...) #copy from drop_read_csv
}
HSMM_epi_3<-drop_read_RDS(paste0(outputDir,"HSMM_epi_3"), dtoken=token)

#How to setup upload UI

#---------2017/6/14 Extract data from S4 and rebuild DESeq------------
"simpler than I thought"
"while running estimateSizeFactors, the error keeps coming:
Error in estimateSizeFactorsForMatrix(counts(object), locfunc = locfunc,  :
every gene contains at least one zero, cannot compute log geometric means
"
#run this script:
cts <- counts(ddsHTSeq)
geoMeans <- apply(cts, 1, function(row) exp(sum(log(row[row != 0]))/length(row)))
dds <- estimateSizeFactors(ddsHTSeq, geoMeans=geoMeans)

#---------2017/6/15 multiple level of DESeq design------------
#https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#contrasts


res_cluster_2vs1 <- results(dds,contrast = c("Cluster","2","1"), alpha=0.5, lfcThreshold=1)
res_cluster_3vs1 <- results(dds,contrast = c("Cluster","3","1"), alpha=0.5, lfcThreshold=1)
res_cluster_3vs2 <- results(dds,contrast = c("Cluster","3","2"), alpha=0.5, lfcThreshold=1)

#Some useful function for S4 vector
isS4()
class?ExpressionSet
HSMM_epi_3@phenoData =phenoData(HSMM_epi_3)
new_object <- updateObject(old_object)
showMethods("as.data.frame")
getMethod("as.data.frame", "DataFrame")
method?as.data.frame
DataFrame


#---------2017/6/16 dds vs dds_results------------
#generate plotCounts plot 1:9 ggplot
d<-data.frame(x= str(0), y= integer(0))
for( i in 1:9){
d<-plot(,returnData=T)
d<- rbind(d,d1)
}
ggplot(d)+facet_warp(~gene_short_name)

#facet_warp vs facet_grid

#generate Rshiny
"print(p), but forgot p<-ggplot"

#"Before publish to shiny, restart the R to double check the package dependency

#//
#//  Debug.R
#//  
#//
#//  Created by Yang Hu on 2017/6/20.
#//
#//

#include <stdio.h>

sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)
#Error in datk[c(startG:endG), ] = foreach(t = actualThreads, .combine = rbind) %dopar%  :
#number of items to replace is not a multiple of replacement length
#"The code below uses parallel computation where multiple cores are available. This works well when R is run from a terminal or from the Graphical User Interface (GUI) shipped with R itself, but at present it does not work with RStudio and possibly other third-party R environments." -WGCNA package tutorial




#Import data into R with an unknown number of columns?
#You need to use the fill argument in the read.table function
fill=TRUE
#There is nice function count.fields (see help) which counts number of column per row:
no_col <- max(count.fields("test", sep = "\t"))
data <- read.table("test",sep="\t",fill=TRUE,col.names=1:no_col)
data

#remove header which have "!"
#grepl returns a logical vector
grepl

#
grep("!Sample_title",MicroArrayData0[,1])
MicroArrayData0[,1]=="!Sample_title"



#the factor variable retains all of its original levels -- even when they do not exist in the new data frame.
droplevels

#Error in sort.list(y) : 'x' must be atomic for 'sort.list'
#Have you called 'sort' on a list?
unlist

#split column into multiple ones
str_split_fixed

#remove rows with empty value in column 1
X <- X[!X[,1]=="",]

#Increase momery
memory.limit(size=65000)

#---2017-07-01
"What is the difference between gsub and sub methods in r"
#The g stands for global, as in replace globally (all):

sub('l', '*', "hello")  #=> "he*lo"
gsub('l', '*',"hello")  #=> "he**o"

#sub(".*") can remove undefined content
filename <- c("GSM726928_OOC1.CEL.gz","GSM726949_STC2_1.CEL.gz")
sampleNames <- sub(".*_", "", filename) #=> "OOC1.CEL.gz" "1.CEL.gz"
sampleNames <- sub(".CEL.gz$", "", sampleNames) #=> "OOC1" "1"


#Avoid rbind()/cbind() conversion from numeric to factor
cbind.data.frame

#merge unequal dataframes and replace missing rows with 0
merge(df1, df2, all = TRUE)

#adjust all log data to positive
logcounts0 <- cpm(expr0,log=TRUE)
min(logcounts0)
all((logcounts0+5.2)>0)

#What does %>% function mean in R?
#The dplyr package introduced the %.% operator to pass the left hand side
#as an argument of the function on the right hand side, similar to a *NIX pipe. |
df <- data.frame(
        x = sample(10, 10, rep = TRUE),
        y = sample(10, 10, rep = TRUE),
        z = sample(10, 10, rep = TRUE)
)
nrow(df)
nrow(distinct(df))
nrow(distinct(df, x, y))
nrow(distinct(df, x))

distinct(df, x)
distinct(df, y)

# Can choose to keep all other variables as well
distinct(df, x, .keep_all = TRUE)
distinct(df, y, .keep_all = TRUE)

# You can also use distinct on computed variables
distinct(df, diff = abs(x - y))

# The same behaviour applies for grouped data frames
# except that the grouping variables are always included
df <- tibble(
        g = c(1, 1, 2, 2),
        x = c(1, 1, 2, 1)
) %>% group_by(g)
df %>% distinct()
df %>% distinct(x)


#To remove the subtitle use the following:
plot(hc, xlab="", sub="")

#----------Task on 2017/07/05----------------
#try to remove genes that don't' change much in the RNAseq experiment
thresh <- myCPM > 0.5 #thresh number decide lower limits of std
keep <- rowSums(thresh) >= 9 # keep rowsum number decide individule circles

#make a heatmap out of the correlation matrix c:
#which microarray sample is each of Danwei's experiments closest to
pheatmap(cor(rnaseq_MicroArrayData, method="spearman"))
#pickup gene set with p value <0.05
#Do deseq, 

#find out linkID
#complete WGCNA


#Plot a data frame as a table in r
library(gridExtra)
title <-as.data.frame(eData$title)
dev.off()
grid.table(title)

# =====================================================================================
# 
#  4. DEseq  ----some values in assay are not integers
# 
# =====================================================================================
# read count matrix
rnaseq_MicroArrayData <-read.csv("rnaseq_MicroArrayData.csv", row.names = 1)

# create sample table for DESeq
sample.table <- data.frame( title = colnames(rnaseq_MicroArrayData),
                            group = c(rep("Danwei",9),
                                      as.character(eData$labelDescription))
)
dim(sample.table) 
## Creating a DESeqDataSet object
#create ddsHTSeq
ddsHTSeq <- DESeqDataSetFromMatrix( countData = rnaseq_MicroArrayData,
                                    colData = sample.table,
                                    design= ~ group)
ddsHTSeq


#Combining data frames with unequal number of columns [duplicate]
df1 = data.frame(A = c(ncol(rnaseq_MicroArrayData),2,1))
df2 = data.frame(B = c("#","Danwei","StemCell","StemCellLine"))
df3 = data.frame(C=c(rep("Danwei",9),as.character(eData$labelDescription)))
l <- merge(t(df1),t(df2),all=TRUE)
l <- merge(l,t(df3),all=TRUE)


#----------Task on 2017/07/07
#try to keep the genes that have higher std (top 25%) in affrymetric data
sd = apply(rnaseq_MicroArrayData[,10:ncol(rnaseq_MicroArrayData)], 1, sd)
sd = sd[order(sd, decreasing=TRUE)]
Top_Std_genes <-names(sd)[1:(length(sd)*0.25)]

# The las argument rotates the axis names
par(las=2)

#similarity of signatures
#find out linkID
#complete WGCNA

#----------Task on 2017/07/08
#add trend line in plot r
lines(lowess(median,sd), col="blue")


#generate table in R markdown
cellType.table <- read.delim("CellType.txt")
kable(cellType.table, captioin ="Cell Type")


#generate table in R
x <- data.frame(row.names=paste("Name",1:10))
x[,1] <- 1:10
x[,2] <- sample(1:100,10)
x[,3] <- sample(LETTERS[1:26],10)
colnames(x) <- c("Value 1", "Value 2", "Label")

#Plot your table with table Grob in the library(gridExtra)
ss <- tableGrob(x)

#Make a scatterplot of your data
k <- ggplot(x,aes(x=x$"Value 1",y=x$"Value 2")) + 
        geom_point()

#Arrange them as you want with grid.arrange
grid.arrange(ss)

#2017-07-14
#try NK cells and other myloide cells

#CD3G(-),NCAM1(+),KLRD1 (+),NCR1(+)	, CD3-, CD56+, CD94+, NKp46+

#assaign each cluster to cell type

#2017-07-18
#FlowJo pineer in single cell level clustering using machine learning
#k-means clustering, tsne 
#http://www.the-scientist.com/?articles.view/articleNo/49228/title/Picking-Out-Patterns/
#https://followthedata.wordpress.com/2015/12/21/list-of-deep-learning-implementations-in-biology/
ADAGE #Analysis using Denoising Autoencoders for Gene Expression 
#https://github.com/greenelab/adage

#2017-07-20
#counts non-zero data
apply(exprs(gbm_cds), 2, function(c)sum(c!=0)) 

#ggplot range
g+coord_cartesian(xlim = c(0, 30000))
g+coord_cartesian(xlim = c(0, 10000))

#density ggplot
qplot(counts_per_cell, data=pData(gbm), geom="density")

#2017-07-21
#lazy-load database 'C:/Users/User/Documents/R/win-library/3.4/VGAM/R/VGAM.rdb' is corrupt
remove.packages("VGAM")
install.packages("VGAM")

#`maximal number of DLLs reached.
#stop loading large amount of libraries

#ggplot
#http://ggplot2.tidyverse.org/reference/index.html#section-plot-basics

#2017-07-24
#gbm for counts
#sce for fpkm

#Error: Aesthetics must be either length 1 or the same as the data (1397): x, y

#2017-07-30-----------------
"how to store variable to list in for loop?"
#list all files with a specified extension
tab.files <-list.files(pattern = "\\.tab$") 
#creat a empty list
tabfileslist <- list() 
for(i in 1:n){
        tab <- tab.files[i] #get file name
        #store variable to list
        tabfileslist[[tab]] <- read.delim(tab) 
        #select 100 genes and 4 columns
        tabfileslist[[tab]] <- tabfileslist[[tab]][1:100,select_columns]
        #filter out genes with RNA.TS score not greater than 0
        tabfileslist[[tab]] <-subset(tabfileslist[[tab]],tabfileslist[[tab]][,"RNA.TS"]>0)
        print(dim(tabfileslist[[tab]]))
}

"How to trim multiple strings"
celltype <- gsub('.tab', '',tab.files)  #"B_Cells.tab" => "B_Cells"

"How to make heatmap with groupedrows and columns"
#https://stackoverflow.com/questions/29740957/how-to-create-pre-annotated-rowside-column-in-heatmap-2
#https://stackoverflow.com/questions/41648293/r-pheatmap-perform-clustering-and-show-dendrograms-per-annotation-category

        "1). pheatmap doesn't work"
#skip all Arguments and add back one by one
#delete arguments filename = "TEST_12cat.png"....

        "2).R list to data frame without lossing any information"
do.call(rbind.data.frame, tabfileslist)
        
        "3). Make a column with duplicated values unique in a dataframe"
make.unique(as.character(myexprs$Gene), sep = ".")

#2017-08-01
"Wireless is slowing down"
#download DNS jumper and chose the DNS with loweast ping

"How to make heatmap with groupedrows and columns for ambigous cell type"
        
        "4).insert a column to the 1st of column of data.frame"
        data.frame$A <- data # works
        matrix$A <-data # doens't work
        data.frame <- data.frame[,c("A",colnames(data.frame))] #switch column A to the 1st 
        
        "5).Make matrix with duplicated Ensembl_exprs row"
        Gene_Ensembl_exprs <- dplyr::left_join(Gene_Ensembl, Ensembl_exprs,by = "Ensembl") #didn't use inner_join, need to name the rows
        
        "6).remove duplicated gene name with lower priority number"
        Gene_Ensembl_celltype[,!duplicated(Gene_Ensembl_celltype$Gene)] # doesn't return the 1st duplicates
        
        "7).order()"
        #be aware of class when order()
        A <-c(4,5,6,7,8, 10,20,30,47)
        B <-as.character(A)
        A[order(A)] #4  5  6  7  8 10 20 30 47
        B[order(B)] #"10" "20" "30" "4"  "47" "5"  "6"  "7"  "8" 
        
#2017-08-04
"use heatmap to show correlation between scRNA and pure cell type"
        "1). replace probe ID with ensembl ID "

"make spearman correlation for pure immue cell type"

        "1). read.table() error 'line 2 did not have 2 elements'"
        #replace space with _, especially for column names


        "2). read.table() error 'more columns than column names'" 
        'as.character(ImmuneCell_raw[1,]), it shows:
       [1] "55805" "18695" "18707" "..."'
        # instead using as.character(ImmuneCell_raw[1,])
        #use
        sapply(ImmuneCell_raw[1,],as.character)
        
        '3).Error in cor(ImmuneCell, method = "spearman") : x must be numeric'
        "Error: (list) object cannot be coerced to type 'double'"
        # http://emilkirkegaard.dk/en/?p=5412
        as.matrix(as.data.frame(lapply(ImmuneCell, as.numeric))) # doesn't work
        
        ImmuneCell.char <- as.matrix(ImmuneCell) #convert to character
        ImmuneCell.matrix <- apply(ImmuneCell.char,2,as.numeric)

"make spearman correlation for pure immue cell type and single cell data"
        "1).no applicable method for 'left_join' applied to an object of class,
        "c('matrix', 'double', 'numeric')""
        SingleCell_ImmuneCell.char <- as.matrix(SingleCell_ImmuneCell)
        SingleCell_ImmuneCell.matrix <- apply(SingleCell_ImmuneCell.char,2,as.numeric)
        
        "2).length vs dim"
        length(data.frame)==ncol(data.frame)
        length(matrix)==nrow(matrix)*ncol(matrix)

"make spearman correlation for pure tissue cell type and single cell data"
        "1)how to subset ExpressionSet object (eData_tissue)"
        select_samples <- sampleNames(eData_tissue) %in% select_samples
        selected_tissue <- eData_tissue[,select_samples]
        
        "2).read.csv blank fields to NA"
        "line 55806 did not have 3 elements"
        fill = TRUE
        
#2017-08-07
        "3).remap Gene.symbol to ensembl"
        "4).remove grid in pheatmap"
        #zoom out
        
"make grouped pheatmap"
        "1).How to extract vectors from multiple S4 objects"
        group <- lapply(eData_names$names,function(x) pData(x)$group)
        group <- unlist(group)
        
        "2).How to extract exprs from multiple S4 objects"
        
        All.exprs <- lapply(eData_names$names,function(x) exprs(x))
        All.exprs <- do.call(cbind,All.exprs) #cbind list
        
        "3).How to creat expressionset"
        samples <- data.frame(title = title,group = group)
        rownames(samples) <-sampleNames
        pd <- new("AnnotatedDataFrame",data=samples)
        tissue_new <- new("ExpressionSet", exprs = All.exprs, 
                          phenoData = pd,
                          annotation = annotation(eData_tissue))
        
#Find the maximum position for each col of a matrix
Max.coe <-max.col(t(c[tissue_Immune_row,n_col]))

#delete multiple digits in vector. FAIL!
?regex
sub(".[[:digit:]]$", "", cell.cluster$Monocle) #delete multiple digits

#2017-08-08
"ggplot trick" 
#http://www.cookbook-r.com/Graphs
#http://www.cookbook-r.com/Graphs/Shapes_and_line_types/

g <-ggplot(cell.cluster, aes(x=Monocle, y=Spearman,
                             color=Spearman.Results, #group color
                             shape=cross.validation))#group shape
g+geom_point(position=position_jitter(width=0.2,height=.2),
             size = 4)+ #"change point size
        ggtitle("Monocle vs Spearman")+#ggplot title
        theme(text = element_text(size=20),     #larger text including legend title							
              legend.position="none", #change legend position
              plot.title = element_text(hjust = 0.5), #title in middle
              axis.text.x  = element_text(angle=70, vjust=0.5))+#rotate xlab
        guides(colour = guide_legend(override.aes = list(size=10)), #larger legend diagram 
               shape = guide_legend(override.aes = list(size=10)))+ #larger legend diagram 
        scale_shape_manual(values=c(19,2,4,1,17,4)) # add manual shape

#2017-08-09
boxplot(eData, las = 3) # rotate xlab


#2017-8-15
TOMsimilarityFromExpr #take long time, have to minimize gene number


#2017-08-18

# Display the dimensions of dataframe in the list
lapply(multiExpr, lapply, dim)
"WGCNA" #Xue nature 2013
"A.To construct the Weighted gene co-expression networks separately"
"1.generated similarity matrices S based on Pearson correlations between all gene pairs"
#Their component Sij was measured by the function, Sij = 0.5 + 0.5 × cor(i,j)., 
#where cor(i,j) denotes Pearson correlation for expression profiles of gene i and j.
adjacency.fromSimilarity
"2.raised the correlation matrix to a user-defined power"
#in order to suppress noisy small correlations and then calculated the adjacency matrices.
"3.calculated topological overlaps"
#which reflect not only direct adjacency in gene pairs but also their interconnections
"4.build hierarchical trees with average linkage method."
#We divided the trees with the Dynamic Tree Cut algorithm to determine gene modules (Langfelder et al., 2008).
"5.merged modules when their eigengenes show strong correlation"

"B.To construct consensus modules"
"1.built adjacency matrices for each dataset using a soft power threshold of 60"
"2.used the adjacency matrices to calculate topological overlap matrices (TOM)"
"3.determined the consensus TOM based on the 25% quantile of multiple TOMs."
"4.The consensus TO was used to build the hierarchical tree, and gene modules were calculated using the Dynamic Tree Cut algorithm."
"5.gene modules were merged when module eigengenes showed a correlation of 75% or better"

"C.Module preservation analysis"
"1-1. examined overlaps between modules derived from different datasets"
"1-2. estimated their statistical significance from the hypergeometric distribution"
"2-1. utilized a composite preservation statistics proposed by WGCNA"
"2-2. calculate Zsummary scores"

#2017-08-19
'Error in numbers2colors(datTraits[, "Cell.Stage"], signed = FALSE) : 
  x must be numeric. For a factor, please use as.numeric(x) in the call.'
#convert character to numeric in r
options(stringsAsFactors = TRUE)
as.numeric(as.factor(datTraits[,"Cell.Stage"]))
test <- data.frame(A =as.numeric(as.factor(datTraits[,"Cell.Stage"])),
                   B =datTraits[,"Cell.Stage"])
test

"get an error when running parallel calculations"
if (!is.na(Sys.getenv("RSTUDIO", unset = NA))) {
        # RStudio specific code
}

#switch function
"EXPR is not a character string it is coerced to integer"

"R switch statement on comparisons"
#findInterval
cuts <- c(0, 20, 30, 40,60, Inf)
chose.power <- c(20, 18,16,14,12)
power <- chose.power[findInterval(nrow(datExpr), cuts)]
power


#5.3.5 Construction of a weighted gene co-expression network and network modules
"if gene <5000, scaleFreePlot shows it doesn't fit the scale free topology"
#5.1.2c Dealing with large datasets: block-wise network construction and module detection

cor(x, y = NULL, use = "everything",
    method = c("pearson", "kendall", "spearman"))
#http://www.statisticssolutions.com/correlation-pearson-kendall-spearman/
#Pearson r correlation :        normally distributed (bell-shaped curve).  
#                               linearity (straight line)
#                               homoscedasticity (normally distributed about the regression line)
#                       Data that is interval or ratio level

#Kendall rank correlation:      strength of dependence between two variables

#Spearman rank correlation:     degree of association between two variables
#                               does not make any assumptions about the distribution

#Spearman rho correlation:      data must be at least ordinal and scores on one variable
#                               must be montonically related to the other variable.
#                       Although this scale allows us to determine greater than, less than, or equal to,
#                       it still does not define the magnitude of the relationship between units. 

"Pareto distribution"  #wiki
"power-law probability distribution" #wiki

"What is the difference between cor and cor.test in R"
#use arguments

#convert character to numeric in r
"Error in sort.list(y) : 'x' must be atomic for 'sort.list'
Have you called 'sort' on a list?"
options(stringsAsFactors = TRUE)
allTraits = apply(traitData,2,function(x) as.numeric(as.factor(x)))
allTraits <- as.data.frame(allTraits)
rownames(allTraits) <- traitData$Cell
options(stringsAsFactors = FALSE)

#Remove columns from dataframe where some of values are NA
moduleTraitCor[ , colSums(is.na(moduleTraitCor)) == 0]
moduleTraitCor[ , apply(moduleTraitCor, 2, function(x) !any(is.na(x)))]

#2017-08-20
"how to use biomAT"

library(biomaRt)
# downlaod annotationfrom biomAT
mart <- useMart("ensembl")
ensembl <- useDataset("hsapiens_gene_ensembl", mart)
annot <-getBM(attributes = c("ensembl_gene_id", #Gene stable ID
                     "external_gene_name", #Gene name
                     "go_id",   #GO term accession
                     "affy_hugene_1_0_st_v1"), #AFFY HuGene 1 0 st v1
#      filters = "external_gene_name", values = allgenes,  #save time with filter
      mart = ensembl)

#Remove all NA values from a vector
d <- d[!is.na(d)]
#Remove the empty element in the vector
d <- d[d !=""] 


#2017-08-21
"Error: ggplot2 doesn't know how to deal with data of class numeric"
#Don't subset data inside qplot

#ggplot
#http://ggplot2.tidyverse.org/reference/index.html#section-plot-basics

+theme_bw() #ggplot white background, but can't change font size

ggplot(data=sale.data,aes(amount))+
        geom_histogram(aes(y=..density..),      # Histogram with density instead of count on y-axis
                       binwidth=70,
                       colour="black", fill="white") +
        ggtitle("Amount per transaction")+
        geom_density(alpha=.2, fill="#CC6666")+
        facet_grid(region ~ .)+
        theme(axis.text = element_text(size = 20),
              axis.text.y=element_blank(), #remove y axis
              axis.ticks.y=element_blank())+#remove y axis
        #        scale_x_discrete(name ="Amount")+ #change x axis lab
        scale_y_continuous(name="Percentage(%)") #change y axis lab

#polynomial vs numeric

"Error in combine_vars(data, params$plot_env, rows, drop = params$drop)"

"Cheat Sheets for Plotting Symbols and Color Palettes"
#http://vis.supstat.com/2013/04/plotting-symbols-and-color-palettes/